File: frontend/src/pages/api/auth/utils/mpc.ts
import { Account } from "near-api-js";
import {
  FinalExecutionOutcome,
  FinalExecutionStatus,
} from "near-api-js/lib/providers";
import bs58 from "bs58";

export interface MpcSignature {
  big_r: {
    affine_point: string;
  };
  s: {
    scalar: string;
  };
  recovery_id: number;
}

export type Secp256k1Signature = `secp256k1:${string}`;

export async function requestMpcSignature(payload: {
  signerAccount: Account;
  contractId: string;
  methodName: string;
  args: any;
}): Promise<MpcSignature> {
  const { signerAccount, contractId, methodName, args } = payload;
  const extractBase64Signature = async (
    txn: FinalExecutionOutcome
  ): Promise<string> => {
    if (
      txn.status &&
      typeof txn.status === "object" &&
      "SuccessValue" in txn.status
    ) {
      return txn.status.SuccessValue as string;
    }
    throw new Error("Failed to extract signature from transaction outcome.");
  };

  try {
    const promise = await signerAccount.functionCall({
      contractId,
      methodName,
      args,
      gas: BigInt("300000000000000"),
      attachedDeposit: BigInt("0"),
    });

    const base64Signature = await extractBase64Signature(promise);
    // Decode from base64 -> JSON -> Vec<Option<MPCSignature>>
    const rawBuffer = Buffer.from(base64Signature!, "base64");
    const parsed = JSON.parse(rawBuffer.toString()) as MpcSignature;
    console.log("MPC Signatures:", parsed);
    return parsed;
  } catch (e: any) {
    if (e.context?.transactionHash) {
      const txHash = e.context.transactionHash;

      let signatures: MpcSignature | undefined = undefined;
      try {
        const finalResult = await waitForTransaction(
          signerAccount.connection.provider,
          txHash,
          contractId,
          400000,
          3000
        );

        const base64Signature = await extractBase64Signature(finalResult);
        // Decode from base64 -> JSON -> Vec<Option<MPCSignature>>
        const rawBuffer = Buffer.from(base64Signature!, "base64");
        const parsed = JSON.parse(rawBuffer.toString()) as MpcSignature;
        console.log("MPC Signatures:", parsed);
        return parsed;
      } catch (pollError: any) {
        console.error(`Polling failed: ${pollError.message}`);
      }

      if (!signatures) {
        console.error("No signatures found in the final transaction.");
        throw new Error("No signatures found in the final transaction.");
      }

      return signatures;
    } else {
      console.error("Unexpected error during function call:", e);
      throw e;
    }
  }

  return [];
}

export async function waitForTransaction(
  provider: any,
  txHash: string,
  contractId: string,
  totalTimeout: number = 40000, // 40 seconds
  pollInterval: number = 3000 // 3 seconds
): Promise<FinalExecutionOutcome> {
  const startTime = Date.now();

  while (Date.now() - startTime < totalTimeout) {
    try {
      const result: FinalExecutionOutcome = await provider.txStatus(
        txHash,
        contractId,
        "FINAL"
      );

      // Type guard: Check if status is FinalExecutionStatus (object)
      if (typeof result.status === "object") {
        const status = result.status as FinalExecutionStatus;

        if ("SuccessValue" in status) {
          return result;
        } else if ("Failure" in status) {
          throw new Error(
            `Transaction failed: ${JSON.stringify(status.Failure)}`
          );
        }
      } else {
        // status is FinalExecutionStatusBasic (enum)
        console.log(`Transaction status: ${result.status}`);
      }
    } catch (error: any) {}

    await sleep(pollInterval);
  }

  throw new Error("Transaction polling timed out after 40 seconds.");
}

export async function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => {
    setTimeout(resolve, ms);
  });
}

export function convertMpcSignatureToSecp256k1(
  sig: MpcSignature
): Secp256k1Signature {
  // Helper to remove a "0x" prefix if present
  const strip0x = (hex: string) => (hex.startsWith("0x") ? hex.slice(2) : hex);

  // Get the r value and remove the compressed point prefix (first byte) if present
  let r = strip0x(sig.big_r.affine_point);
  if (r.length === 66) {
    // 66 hex chars = 33 bytes
    r = r.slice(2); // remove the first byte (prefix)
  }
  r = r.padStart(64, "0"); // Ensure it's 32 bytes (64 hex characters)

  // Process s (ensure no "0x" and pad to 64 characters)
  const s = strip0x(sig.s.scalar).padStart(64, "0");

  // Convert recovery_id to a 2-character hex string (1 byte)
  const recoveryId = sig.recovery_id.toString(16).padStart(2, "0");

  // Correct concatenation: r + s + recoveryId
  const hexCombined = r + s + recoveryId;

  // Convert hex string to a byte array
  const bytes = Buffer.from(hexCombined, "hex");

  // Encode the bytes in base58
  const base58Combined = bs58.encode(bytes);

  // Return with the proper prefix
  return `secp256k1:${base58Combined}`;
}

File: agent/src/src/mpc.py
import base64
import base58
import json
import asyncio
import time
from py_near.models import TransactionResult 
from py_near.account import Account

from near_api.account import Account
from near_api.providers import JsonProvider

import traceback



class MpcSignature:
    def __init__(self, big_r: dict, s: dict, recovery_id: int):
        self.big_r = big_r
        self.s = s
        self.recovery_id = recovery_id


Secp256k1Signature = str


async def request_mpc_signature(payload: dict) -> MpcSignature:
    signer_account: Account = payload["signer_account"]
    contract_id = payload["contract_id"]
    method_name = payload["method_name"]
    args = payload["args"]

    async def extract_base64_signature(txn: TransactionResult) -> str:
        try:
            status = txn.status
            if status and isinstance(status, dict) and 'SuccessValue' in status:
                return status['SuccessValue']
        except Exception as e:
            print(f"excpetion: {e}")
            traceback.print_exc()
            raise Exception("Failed to extract signature from transaction outcome.")

    try:
        print(f"args: {args}")
        promise: TransactionResult = await signer_account.call(
            contract_id=contract_id,
            method_name=method_name,
            args=args,
            gas=300000000000000,
            amount=0,
        )
        print(f"balance promise: {promise.status}")
        base64_signature = await extract_base64_signature(promise)
        raw_buffer = base64.b64decode(base64_signature)
        parsed = json.loads(raw_buffer.decode())  # Assuming the response is a MpcSignature
        print("MPC Signatures:", parsed)
        return MpcSignature(**parsed)

    except Exception as e:
        print(f"Rest exception: {e}")
        if hasattr(e, "context") and hasattr(e.context, "transaction_hash"):
            tx_hash = e.context.transaction_hash
            
            try:
                final_result = await wait_for_transaction(
                    signer_account.provider(),
                    tx_hash,
                    contract_id,
                    total_timeout=400000,
                    poll_interval=3000,
                )

                base64_signature = await extract_base64_signature(final_result)
                raw_buffer = base64.b64decode(base64_signature)
                parsed = json.loads(raw_buffer.decode())  # Assuming the response is a MpcSignature
                print("MPC Signatures:", parsed)
                return MpcSignature(**parsed)

            except Exception as poll_error:
                print(f"Polling failed: {poll_error}")

            raise Exception("No signatures found in the final transaction.")

        else:
            print("Unexpected error during function call:", e)
            raise e


async def wait_for_transaction(
    provider,
    tx_hash: str,
    contract_id: str,
    total_timeout: int = 40000,
    poll_interval: int = 3000,
) :
    start_time = time.time()

    while (time.time() - start_time) * 1000 < total_timeout:
        try:
            result = await provider.get_tx_status(tx_hash, contract_id)
            print(f"Result: {result}")

            if isinstance(result.status, dict):
                status = result.status

                if "SuccessValue" in status:
                    return result
                elif "Failure" in status:
                    raise Exception(f"Transaction failed: {json.dumps(status['Failure'])}")
            else:
                print(f"Transaction status: {result.status}")
        except Exception as e:
            print(e)
            traceback.print_exc()

        await sleep(poll_interval)
    print("Transaction polling timed out after 40 seconds.")
    raise Exception("Transaction polling timed out after 40 seconds.")


async def sleep(ms: int) -> None:
    await asyncio.sleep(ms / 1000)


def convert_mpc_signature_to_secp256k1(sig: MpcSignature) -> Secp256k1Signature:
    def strip_0x(hex_value: str) -> str:
        return hex_value[2:] if hex_value.startswith("0x") else hex_value

    r = strip_0x(sig.big_r["affine_point"])
    if len(r) == 66:
        r = r[2:]  # Remove the first byte (prefix)
    r = r.zfill(64)  # Ensure it's 32 bytes (64 hex characters)

    s = strip_0x(sig.s["scalar"]).zfill(64)

    recovery_id = hex(sig.recovery_id)[2:].zfill(2)

    hex_combined = r + s + recovery_id

    bytes_combined = bytes.fromhex(hex_combined)
    base58_combined = base58.b58encode(bytes_combined).decode()

    return f"secp256k1:{base58_combined}"

